File: java/src/main/java/org/twyszomirski/aoc/solutions/Day_13.java
package org.twyszomirski.aoc.solutions;


import io.vavr.Tuple2;
import io.vavr.collection.Stream;
import org.twyszomirski.aoc.utils.Pair;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;
import java.util.stream.Gatherers;
import java.util.stream.IntStream;

import static java.io.IO.println;

public class Day_13 {

    public static void solve() throws IOException {
        var lines = Files.readAllLines(Paths.get("src/main/resources/input_day_13.txt"));
        println("======== Day 13 ===========");
        part_1(lines);
        part_2(lines);
    }

    private static void part_1(List<String> lines) {
        var games = getData(lines);
        var sum = games.stream().map(Day_13::solve).filter(Objects::nonNull)
                .mapToLong(buttons -> buttons.left() * 3 + buttons.right()).sum();
        println(sum);
    }

    private static void part_2(List<String> lines) {
        var addition = 10000000000000L;
        var games = getData(lines);
        var withAddition = games.stream().map(game -> new Game(game.buttonA, game.buttonB, new Prize(game.prize.x+addition, game.prize.y + addition)));
        var sum = withAddition.map(Day_13::solve).filter(Objects::nonNull)
                .mapToLong(buttons -> buttons.left() * 3 + buttons.right()).sum();
        println(sum);
    }

    private static Pair<Long> solve(Game game) {
        /*
            a * Ax + b* Bx = Prize.x
            a * Ay + b* By = Prize.y
        */

        var w = (game.buttonA.x * game.buttonB.y) - (game.buttonB.x * game.buttonA.y);
        var wX = (game.prize.x * game.buttonB.y) - (game.prize.y * game.buttonB.x);
        var wY = (game.buttonA.x * game.prize.y) - (game.buttonA.y * game.prize.x);
        if (w != 0L) {
            if ((wX % w == 0L && wY % w == 0L)) return new Pair<>((wX / w), (wY / w));
            else return null;
        }
        return null;
    }

    private static List<Game> getData(List<String> lines) {
        return lines.stream().filter(it -> !it.isBlank())
                .gather(Gatherers.windowFixed(3)).map(data ->
                {
                    var a = data.get(0);
                    var b = data.get(1);
                    var p = data.get(2);
                    var buttonA = new Button(Long.parseLong(a.substring(a.indexOf("X+") + 2, a.indexOf(","))), Long.parseLong(a.substring(a.indexOf("Y+") + 2)));
                    var buttonB = new Button(Long.parseLong(b.substring(b.indexOf("X+") + 2, b.indexOf(","))), Long.parseLong(b.substring(b.indexOf("Y+") + 2)));
                    var prize = new Prize(Long.parseLong(p.substring(p.indexOf("X=") + 2, p.indexOf(","))), Long.parseLong(p.substring(p.indexOf("Y=") + 2)));
                    return new Game(buttonA, buttonB, prize);
                }).toList();
    }

    record Button(long x, long y) {
    }

    record Prize(long x, long y) {
    }

    record Game(Button buttonA, Button buttonB, Prize prize) {
    }

}

File: kotlin/src/main/kotlin/org/twyszomirski/aoc/solutions/Day_13.kt
package org.twyszomirski.aoc.solutions

import java.io.File


class Day_13 {

    fun solve() {
        val lines = File("src/main/resources/input_day_13.txt").readLines()
        println("======== Day 13 ===========")
        part1(lines)
        part2(lines)
    }

    fun part1(input: List<String>) {
        val data = readData(input)

        val sum = data.map {
            val onX = findSolutions(game = it, pointAccessor = { button -> button.x }, prizeAccessor = { p -> p.x })
            val onY = findSolutions(game = it, pointAccessor = { button -> button.y }, prizeAccessor = { p -> p.y })
            onX.intersect(onY)
        }.map { it.sortedBy { it.first * 3 + it.second } }
            .mapNotNull { it.firstOrNull() }.sumOf { it.first * 3 + it.second }

        println(sum)
    }

    fun part2(input: List<String>) {
        val data = readData(input)
        val addition = 10000000000000L
        val sum = data
            .map { Game(it.buttonA, it.buttonB, Prize(it.prize.x + addition, it.prize.y + addition)) }
            .mapNotNull {
                findSolutionsByEquation(it)
            }
            .sumOf { it.first * 3 + it.second }

        println(sum)
    }

    fun findSolutions(
        game: Game,
        pointAccessor: (Button) -> Long,
        prizeAccessor: (Prize) -> Long
    ): List<Pair<Long, Long>> {
        val prize = prizeAccessor(game.prize)
        val aValue = pointAccessor(game.buttonA)
        val bValue = pointAccessor(game.buttonB)
        val sol = if (aValue > bValue) {
            findSolutions(aValue, bValue, prize)
        } else {
            findSolutions(bValue, aValue, prize)
                .map { Pair(it.second, it.first) }
        }
        return sol
    }

    fun findSolutions(bigger: Long, lower: Long, result: Long): List<Pair<Long, Long>> {
        var countOfBigger = result / bigger
        val matches = mutableListOf<Pair<Long, Long>>()
        while (countOfBigger > 0) {
            val rest = result - (countOfBigger * bigger)
            val countOfLower = rest / lower
            if (addsUp(bigger, lower, result, countOfBigger, countOfLower)) {
                matches.add(Pair(countOfBigger, countOfLower))
            }
            countOfBigger--
        }
        return matches
    }

    fun findSolutionsByEquation(
        game: Game
    ): Pair<Long, Long>? {
        /*
            a * Ax + b* Bx = Prize.x
            a * Ay + b* By = Prize.y
        */

        val w = (game.buttonA.x * game.buttonB.y) - (game.buttonB.x * game.buttonA.y)
        val wX = (game.prize.x * game.buttonB.y) - (game.prize.y * game.buttonB.x)
        val wY = (game.buttonA.x * game.prize.y) - (game.buttonA.y * game.prize.x)
        if (w != 0L) {
            return if ((wX % w == 0L && wY % w == 0L)) Pair((wX / w), (wY / w)) else null
        }
        return null
    }

    fun addsUp(bigger: Long, lower: Long, result: Long, countOfBigger: Long, countOfLower: Long): Boolean {
        return (bigger * countOfBigger) + (lower * countOfLower) == result
    }


    private fun readData(lines: List<String>): List<Game> {
        return lines.filter { it.isNotBlank() }.windowed(3, 3).map {
            val buttonA = it.get(0)
            val buttonB = it.get(1)
            val prize = it.get(2)

            val a = Button(
                buttonA.substring(buttonA.indexOf("X+") + 2, buttonA.indexOf(",")).toLong(),
                buttonA.substring(buttonA.indexOf("Y+") + 2).toLong())

            val b = Button(
                buttonB.substring(buttonB.indexOf("X+") + 2, buttonB.indexOf(",")).toLong(),
                buttonB.substring(buttonB.indexOf("Y+") + 2).toLong())

            val p = Prize(
                prize.substring(prize.indexOf("X=") + 2, prize.indexOf(",")).toLong(),
                prize.substring(prize.indexOf("Y=") + 2).toLong()
            )

            Game(a, b, p)
        }
    }
}

data class Button(val x: Long, val y: Long)
data class Prize(val x: Long, val y: Long)
data class Game(val buttonA: Button, val buttonB: Button, val prize: Prize)


